target/riscv: Mismatch napot when mcontrol.maskmax is not expected

sunnyzhu-learning opened the pull request:
From https://github.com/riscv-collab/riscv-openocd/pull/1124,
1.Rename `trigger_request_info::tdata1_ignore_mask` -> `trigger_request_info::maskmax`.
2.Add the something like the following to `set_trigger()`:

>  const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
if (type == CSR_TDATA1_TYPE_MCONTROL &&
		(get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT))
		tdata1_config_denied |= get_field(tdata1_rb, CSR_MCONTROL_MASKMAX) < (get_least_significant_zero_index(tdata2) + 1).
}

addressed
@en-sc Thanks for your suggestion, internal discussion a lot,  and first push code to github,delay....please review~

en-sc commented:
Checkpatch fails:
```
> Run ./tools/scripts/checkpatch.pl --no-signoff --git FETCH_HEAD..HEAD
WARNING:COMMIT_MESSAGE: Missing commit description - Add an appropriate one
...
```
Please, add an empty line between commit title and description.
See https://git-scm.com/docs/git-commit#_discussion

en-sc commented on src/target/riscv/riscv.c:
```diff
@@ -619,8 +619,17 @@ static int find_first_trigger_by_id(struct target *target, int unique_id)
 	return -1;
 }
 
+static inline uint64_t get_least_significant_zero_index(riscv_reg_t tdata2)
```
1. Since the definition is not in a header I believe `inline` to be redundant.
2. `unsigned int` seems more appropriate here as a return type:
https://github.com/riscv-collab/riscv-openocd/blob/c85d4e1858840fe90a2fa0c521528338952d49da/doc/manual/style.txt#L168-L169
3. IMHO, `tdata2` is better renamed to`reg` or some other generic name -- there is nothing specific to `tdata2` about this function.
4. If you like: `get_least_significant_zero_index` -> `count_trailing_ones` (seems to me more specific and closer to naming scheme of compiler builtins).
```suggestion
static unsigned int count_trailing_ones(riscv_reg_t reg)
```

sunnyzhu-learning replied:
addressed

en-sc reviewed (commented):
Only minor comments, nothing significant.

@JanMatCodasip, @MarekVCodasip, could you please take a look?

On a side note:
@sunnyzhu-learning, have you tested the change? What was the simulator/HW used?

sunnyzhu-learning commented:
Q:have you tested the change? What was the simulator/HW used?
A:tested in HW env,
following is gdb + OpenOCD test：
```
$ riscv64-unknown-elf-gdb hello_world.elf
Reading symbols from hello_world.elf...
(gdb) target remote:3333
Remote debugging using :3333
_start () at ../****/SoC/****/Common/Source/GCC/startup_***.S:111
111         csrc MSTATUS, MSTATUS_MIE
(gdb) load
Loading section .init, size 0x18e lma 0x80000000
Loading section .text, size 0xe38c lma 0x80000400
Loading section .rodata, size 0xd00 lma 0x8000e790
Loading section .init_array, size 0x4 lma 0x8000f490
Loading section .data, size 0x728 lma 0x800104a0
Loading section .amo, size 0x90 lma 0x80011f58
Start address 0x80000000, load size 63958
Transfer rate: 34 KB/sec, 7106 
(gdb) monitor wp 0x800104a0 0x2 r
[riscv.cpu] Found 4 triggers
[riscv.cpu] Could not set a trigger that will match a whole address range. As a fallback, this trigger (and maybe others) will only match against the first address of the range.
(gdb) monitor reg tselect 0
tselect (/32): 0x00000000
(gdb) monitor reg tdata1
tdata1 (/32): 0x28001049
(gdb) monitor reg tdata2
tdata2 (/32): 0x800104a0

en-sc reviewed (dismissed):
LGTM

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -619,8 +619,17 @@ static int find_first_trigger_by_id(struct target *target, int unique_id)
 
+static unsigned int count_trailing_ones(riscv_reg_t reg)
+{
+	for (int i = 0; i < 64; i++) {
```
Very small detail:

This helper function assumes that `riscv_reg_t` is an alias to `uint64_t`. 

Please, consider putting an assertion there:

```suggestion
	assert(sizeof(riscv_reg_t) * 8 == 64);

	for (int i = 0; i < 64; i++) {
```

sunnyzhu-learning replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -716,13 +731,13 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
 	riscv_reg_t tdata1;
 	riscv_reg_t tdata2;
-	riscv_reg_t tdata1_ignore_mask;
+	riscv_reg_t maskmax;
```
I am not in favor of this rename. It can mis-lead the code reader to think that this field contains a value to be written to `maskmax` field, which is however not true.

Please, consider one of the following:

- a) Keep the original name `tdata1_ignore_mask` which very well describes the meaning of this variable.
- b) Replace this field by a simple `bool` flag that says whether the special handling of maxmask should be activated or not. E.g.: `bool check_tdata1_maxmask;`

Thank you.

en-sc replied:
Maybe dropping this field altogether is better?
`maskmax`/`tdata1_ignore_mask` is used when `tdata1.type` is `mcontrol` (2).
This commit already derives `type` from `tdata1` in `set_trigger()`.
https://github.com/riscv-collab/riscv-openocd/blob/f0bbdb37724f2327bce6f4603273266a02a1d9b6/src/target/riscv/riscv.c#L657

Wouldn't ignoring `mcontrol.maskmax` based on `tdat1.type` be better:
```
	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
	const riscv_reg_t tdata1_ignore_mask = (type == CSR_TDATA1_TYPE_MCONTROL) ?
			CSR_MCONTROL_MASKMAX(riscv_xlen(target)) : 0;
	bool tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);
	if (type == CSR_TDATA1_TYPE_MCONTROL &&
	    (get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT)) {
	    tdata1_config_denied |= (get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)))
				    < (count_trailing_ones(tdata2) + 1));
	}
```
@JanMatCodasip, what are your thoughts on this?

JanMatCodasip replied:
I like that solution :+1:. It is probably better to go that way, compared to what I originally suggested.

sunnyzhu-learning replied:
I have dropped maskmax,this value derives in `set_trigger()`

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -644,15 +653,21 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+				tdata1, tdata1_rb, maskmax);
 
 		if (tdata2_config_denied)
 			LOG_TARGET_DEBUG(target,
```
Detail but helpful:

While making changes there, it would be good to unify the debug messages for tdata1 and tdata2:

```
	if (tdata2_config_denied)
			LOG_TARGET_DEBUG(target,
				"After writing 0x%" PRIx64 " to tdata2 it contains 0x%" PRIx64,
				tdata2, tdata2_rb);
```

sunnyzhu-learning replied:
addressed

JanMatCodasip replied:
Thanks.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -644,15 +653,21 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
 	if (riscv_reg_get(target, &tdata2_rb, GDB_REGNO_TDATA2) != ERROR_OK)
 		return ERROR_FAIL;
-	bool tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);
+	bool tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	if (type == CSR_TDATA1_TYPE_MCONTROL &&
+	    (get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT)) {
+	    tdata1_config_denied |= (get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)))
+				    < (count_trailing_ones(tdata2) + 1));
+	}
```
1) A third boolean flag for should be introduced here, e.g. `unsupported_napot_range`. The existing flag `tdata1_config_denied should` should not be re-used because these are two different error situations. There also should be different log messages for each (please see below).

2) Also the type of the trigger (line 657) should be taken from `tdata1_rb`. That's because the target can decide to disable the trigger (by changing the trigger type to `15`), and in that case we cannot get any of the `mcontrol` fields (`mcontrol.match`, `mcontrol.maskmax`) - all those checks become irrelevant.

3) Since the conditions are quite complex, you can improve the code readability a lot by creating helper variables like `is_mcontrol`, `is_napot_match`, ...

```suggestion
	bool tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
	
	bool unsupported_napot_range = false;
	const uint32_t type = get_field(tdata1_rb, CSR_TDATA1_TYPE(riscv_xlen(target)));
	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
	if (is_mcontrol && is_napot_match) {
		const riscv_reg_t maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target));
		const unsigned napot_size = (count_trailing_ones(tdata2) + 1);
		if (maskmax < napot_size) {
			unsupported_napot_range = true;			
		}
	}
```

Error message for `unsupported_napot_range`:

```c
if (tdata1_config_denied || tdata2_config_denied || unsupported_napot_range) {
	// ...

	if (unsupported_napot_range)
		LOG_TARGET_DEBUG(target,
			"The requested NAPOT match range (tdata2=0x%" PRIx64 ") exceeds maxmask=0x%" PRIx64,
			... );
	// ...
	return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
}
```

sunnyzhu-learning replied:
addressed

JanMatCodasip reviewed (changes requested):
Hi @sunnyzhu-learning,

thanks for your contribution! Overall, I like the improvement.

I do have several review comments, though. If you can please look at them.

Thank you.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -619,8 +619,17 @@ static int find_first_trigger_by_id(struct target *target, int unique_id)
 
+static unsigned int count_trailing_ones(riscv_reg_t reg)
+{
+	for (int i = 0; i < 64; i++) {
```
Small detail: It would be good to make the data type match the return type of the function.

```suggestion
	for (unsigned int i = 0; i < 64; i++) {
```

sunnyzhu-learning replied:
addressed

sunnyzhu-learning commented:
Q:what is the reason for closing the PR?
Please, consider reopening. I believe the patch to be very valuable.

A:my branch updated,this PR deleted,and not reopen again ,so I recreated a PR(https://github.com/riscv-collab/riscv-openocd/pull/1133) for this
![image](https://github.com/user-attachments/assets/0f97bc69-c061-4dd2-bb30-069c55b5628d)

Commit:
target/riscv: Mismatch napot when mcontrol.maskmax is not expected

1.Remove trigger_request_info::tdata1_ignore_mask
2.Adding ignore napot matching condition

Signed-off-by: Songhe Zhu <zhusonghe@eswincomputing.com>
Reviewed-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>
Reviewed-by: Jan Matyas <jan.matyas@codasip.com>

sunnyzhu-learning commented:
> @sunnyzhu-learning, what is the reason for closing the PR? Please, consider reopening. I believe the patch to be very valuable.

reopen this PR,solved code merge conflict,I'm very sorry,poor use of PR,@en-sc @JanMatCodasip please code review,thanks....

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -633,7 +642,7 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
 	}
 
-	riscv_reg_t tdata1_rb, tdata2_rb;
+	riscv_reg_t tdata1_rb, tdata2_rb, maskmax;
```
Detail:

The variable `maskmax` can be declared very close to the place of the use, which improves code readability.

```suggestion
	riscv_reg_t tdata1_rb, tdata2_rb;
```

sunnyzhu-learning replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	bool tdata1_config_denied = true;
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
+	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
```
This check should use the real value (read back from the register), not the one that we attempted to write there.

```suggestion
	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
```

sunnyzhu-learning replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	bool tdata1_config_denied = true;
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
+	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
```
(same as previous)

```suggestion
	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
```

JanMatCodasip replied:
(same)

```suggestion
	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
```

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	} else {
+		tdata1_config_denied = tdata1 != tdata1_rb;
+	}
 	bool tdata2_config_denied = tdata2 != tdata2_rb;
```
Detail: The use of `const` is helpful for code readers.

```suggestion
	const bool tdata2_config_denied = tdata2 != tdata2_rb;
```

sunnyzhu-learning replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
+	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
+	if (is_mcontrol && is_napot_match) {
+		maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
+		const unsigned int napot_size = count_trailing_ones(tdata2) + 1;
+		if (maskmax < napot_size)
+			unsupported_napot_range = true;
+		tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
```
While refactoring the code, a bug was introduced: There would need to be separate variables for `maskmax_value` and `maskmax_mask`, as shown below.

Instead of just fixing it, please, improve the overall structure of the code as shown in the below comment.

```c
		maskmax_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target));
		maskmax_value = get_field(tdata1_rb, maskmax_mask);
		const unsigned int napot_size = count_trailing_ones(tdata2) + 1;
		if (maskmax_value < napot_size)
			unsupported_napot_range = true;
		tdata1_config_denied = (tdata1 & ~maskmax_mask) != (tdata1_rb & ~maskmax_mask);
```

sunnyzhu-learning replied:
addressed

sunnyzhu-learning replied:
thanks a lot, before:line：677 value `maskmax` is a bug

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
+	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
+	if (is_mcontrol && is_napot_match) {
+		maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
+		const unsigned int napot_size = count_trailing_ones(tdata2) + 1;
+		if (maskmax < napot_size)
+			unsupported_napot_range = true;
+		tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
+	} else if (is_mcontrol && !is_napot_match) {
+		maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
+		tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
+	} else {
+		tdata1_config_denied = tdata1 != tdata1_rb;
+	}
```
This part of the code combines two almost-independent checks:

- check for the TDATA1 value
- check for unsupported mcontrol NAPOT range

For that reason, please, split it into two separate checks. Don't try to merge both the checks into a single if-else cascade. That way it the code will be easier to comprehend, and bugs like the above will hopefully become less likely.

Also please use blank lines to visually separate related blocks of code. 

Thanks.

```c
/* Determine if tdata1 supports what we need.
 * For mcontrol triggers, we don't care about the value in the read-only "maskmax" field. */
const riscv_reg_t tdata1_ignore_mask = is_mcontrol ? CSR_MCONTROL_MASKMAX(riscv_xlen(target)) ? 0;
const riscv_reg_t tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);

/* Determine if tdata1.maxmask is sufficient
 * (only relevant for mcontrol triggers and NAPOT match type) */
riscv_reg_t unsupported_napot_range = false;
if (is_mcontrol && is_napot_match) {
	// ...
}
```

sunnyzhu-learning replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+		if (unsupported_napot_range)
+			LOG_TARGET_DEBUG(target,
+				"The requested NAPOT match range (tdata2=0x%" PRIx64 ") exceeds maskmax=0x%" PRIx64,
+				tdata2, maskmax);
 		if (riscv_reg_set(target, GDB_REGNO_TDATA1, 0) != ERROR_OK)
```
Detail: Blank line can be used here to maintain consistent style. 

```suggestion
				tdata2, maskmax);

		if (riscv_reg_set(target, GDB_REGNO_TDATA1, 0) != ERROR_OK)
```

sunnyzhu-learning replied:
addressed

JanMatCodasip reviewed (changes requested):
@sunnyzhu-learning Thank you for addressing our feedback so far. 

I have several more comments. If you can please look at them.

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,50 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	 */
+	bool unsupported_napot_range = false;
+	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
+	riscv_reg_t maskmax_mask, maskmax_value;
+	if (is_mcontrol && is_napot_match) {
+		maskmax_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target));
+		maskmax_value = get_field(tdata1_rb, maskmax_mask);
```
For safety, please initialize `maskmax_value`.

Variable `maskmax_mask` can be const. Alternatively, since it is now used in a single place only, it can be removed completely. (Please pick which solution you prefer.)

```suggestion
	riscv_reg_t maskmax_value = 0;
	if (is_mcontrol && is_napot_match) {
		const riscv_reg_t maskmax_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target));
		maskmax_value = get_field(tdata1_rb, maskmax_mask);
```
or

```suggestion
	riscv_reg_t maskmax_value = 0;
	if (is_mcontrol && is_napot_match) {
		maskmax_value = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
```

sunnyzhu-learning replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -655,20 +664,50 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
+	 * (only relevant for mcontrol triggers and NAPOT match type)
+	 */
+	bool unsupported_napot_range = false;
+	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
+	riscv_reg_t maskmax_mask, maskmax_value;
+	if (is_mcontrol && is_napot_match) {
+		maskmax_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target));
+		maskmax_value = get_field(tdata1_rb, maskmax_mask);
+		const unsigned int napot_size = count_trailing_ones(tdata2) + 1;
+		if (maskmax_value < napot_size)
+			unsupported_napot_range = true;
+	}
```
One more suggestion: Please wrap the whole maxmask check into `if (!tdata1_config_denied) { ... }`.

Reasons:

- If it is already determined that tdata1 does not support the trigger options that we need, it would be logically incorrect to continue with other checks. 
- Furthermore, `tdata1.type` may even change to something else (e.g. disabled trigger), in which case mcontrol.maskmax field won't even be available for us to read.

```c
	bool unsupported_napot_range = false;
	riscv_reg_t maskmax_value = 0;
	if (!tdata1_config_denied) {
		const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
		if (is_mcontrol && is_napot_match) {
			// ...
		}
	}
```

or

```c
	bool unsupported_napot_range = false;
	riscv_reg_t maskmax_value = 0;
	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
	if (!tdata1_config_denied && is_mcontrol && is_napot_match) {
		// ...
	}
```

(Please pick one of the variants.)

sunnyzhu-learning replied:
Thanks for the explanation, this is a very rigorous implementation, I will wrap whole maskmax in next patch

JanMatCodasip reviewed (commented):
@sunnyzhu-learning Thank you for addressing the review feedback so far.

I have posted two last comments (and that should be all from me).

JanMatCodasip reviewed (approved):
LGTM.

@sunnyzhu-learning, thank you for your contribution.

en-sc commented on src/target/riscv/riscv.c:
```diff
@@ -621,8 +621,17 @@ static int find_first_trigger_by_id(struct target *target, int unique_id)
-	riscv_reg_t tdata1_ignore_mask)
+static unsigned int count_trailing_ones(riscv_reg_t reg)
+{
+	assert(sizeof(riscv_reg_t) * 8 == 64);
+	for (unsigned int i = 0; i < 64; i++) {
```
```suggestion
	const unsigned int riscv_reg_bits = sizeof(riscv_reg_t) * CHAR_BIT;
	for (unsigned int i = 0; i < riscv_reg_bits; i++) {
```

en-sc commented on src/target/riscv/riscv.c:
```diff
@@ -621,8 +621,17 @@ static int find_first_trigger_by_id(struct target *target, int unique_id)
+		if ((1 & (reg >> i)) == 0)
+			return i;
+	}
+	return 64;
```
```suggestion
	return riscv_reg_bits;
```

sunnyzhu-learning commented:
@en-sc  Sorry，No timely feedback，I took a seven-day vacation，next PR will modify above changes~
